mapo/agents/mapo/mapo_policy.py

```diff
@@ -10,23 +12,67 @@
 from ray.rllib.evaluation.postprocessing import compute_advantages, Postprocessing
 
 
-def build_actor_critic_losses(policy, batch_tensors):
-    """Contruct actor (DPG) and critic (Fitted Q) tf losses."""
+AgentComponents = namedtuple("AgentComponents", "dynamics critic actor")
+
+
+def _build_dynamics_mle_loss(batch_tensors, model):
+    obs, actions, next_obs = (
+        batch_tensors[SampleBatch.CUR_OBS],
+        batch_tensors[SampleBatch.ACTIONS],
+        batch_tensors[SampleBatch.NEXT_OBS],
+    )
+    return -tf.reduce_mean(model.compute_states_log_prob(obs, actions, next_obs))
+
+
+def _build_critic_loss(batch_tensors, model):
     # Fitted Q loss (using trajectory returns)
     obs, actions, returns = (
         batch_tensors[SampleBatch.CUR_OBS],
         batch_tensors[SampleBatch.ACTIONS],
         batch_tensors[Postprocessing.ADVANTAGES],
     )
-    policy.dynamics_loss = tf.constant(0.0)  # Placeholder for future losses
-    policy.critic_loss = keras.losses.mean_squared_error(
-        policy.model.compute_q_values(obs, actions), returns
+    return keras.losses.mean_squared_error(
+        model.compute_q_values(obs, actions), returns
     )
-    # DPG loss
-    policy_action = policy.model.compute_actions(obs)
-    policy_action_value = policy.model.compute_q_values(obs, policy_action)
-    policy.actor_loss = -tf.reduce_mean(policy_action_value)
-    return policy.actor_loss + policy.critic_loss + policy.dynamics_loss
+
+
+def _build_actor_loss(batch_tensors, model, config):
+    obs = batch_tensors[SampleBatch.CUR_OBS]
+    gamma = config["gamma"]
+    n_samples = config["branching_factor"]
+    policy_action = model.compute_actions(obs)
+    sampled_next_state, next_state_log_prob = model.compute_log_prob_sampled(
+        obs, policy_action, (n_samples,)
+    )
+    next_state_value = tf.stop_gradient(model.compute_state_values(sampled_next_state))
+    model_aware_policy_loss = tf.reduce_mean(
```

Comment:
BTW.. I added a `_reward_fn` (and also a `_transition_fn`) method in NavigationEnv that takes state and action tensors and return another tensor for the reward computation... I will create a base class to be used for all custom envs that properly implements this API.. but by now we should be able to compute the `mapo_actor_loss` with the reward gradient term... 

Just need to figure it out a way to get access to the environment in the policy... any thoughts?

Reply:
For now, I can think of two ways:
* Add a callback on episode start that calls a custom method (maybe `set_current_env`) to set the environment the policy is using. Callbacks are passed through the config (`config["callbacks"]["on_episode_start"]`), but we can add then in `MAPOTrainer` so that they're added every time ([DQN](https://github.com/ray-project/ray/blob/e4854b11f9a032261a3a07029b8492e59002da61/python/ray/rllib/agents/dqn/dqn.py#L168) does this). The callback is called with the following arguments:
```python
            callbacks["on_episode_start"]({
                "env": base_env,
                "policy": policies,
                "episode": episode,
            })
```
* Create a separate environment in the policy using `config["env"]` and `config["env_config"]`. This can be problematic if the environment has some hidden internal state, since in that case the instance used for calculation transitions and the one used for training might behave differently. Nevertheless, we would probably do something similar to what `Trainer` does:
```python
    @override(Trainable)
    def _setup(self, config):
        env = self._env_id
        if env:
            config["env"] = env
            if _global_registry.contains(ENV_CREATOR, env):
                self.env_creator = _global_registry.get(ENV_CREATOR, env)
            else:
                import gym  # soft dependency
                self.env_creator = lambda env_config: gym.make(env)
        else:
            self.env_creator = lambda env_config: None

```